core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchReaderIterator.java

```diff
@@ -241,38 +247,59 @@ private void binRanges(TabletLocator tabletLocator, List<Range> ranges,
 
     int lastFailureSize = Integer.MAX_VALUE;
 
-    while (true) {
+    if (options.getConsistencyLevel().equals(ConsistencyLevel.EVENTUAL) && !this.tableOnline) {
 
+      // The TabletLocator only caches the location of tablets for online tables. The ScanServer
+      // can scan online and offline tables, but does not require the location of the tablet. It
+      // only requires the KeyExtent. Find the first extent for the table that matches the
+      // startRow provided in the scanState. If no startRow is provided, use the first KeyExtent
+      // for the table
       binnedRanges.clear();
-      List<Range> failures = tabletLocator.binRanges(context, ranges, binnedRanges);
-
-      if (failures.isEmpty()) {
-        break;
-      } else {
-        // tried to only do table state checks when failures.size() == ranges.size(), however this
-        // did
-        // not work because nothing ever invalidated entries in the tabletLocator cache... so even
-        // though
-        // the table was deleted the tablet locator entries for the deleted table were not
-        // cleared... so
-        // need to always do the check when failures occur
-        if (failures.size() >= lastFailureSize) {
-          context.requireNotDeleted(tableId);
-          context.requireNotOffline(tableId, tableName);
+      Map<KeyExtent,List<Range>> map = new HashMap<>();
+      ranges.forEach(r -> {
+        TabletsMetadata tm = context.getAmple().readTablets().forTable(this.tableId)
+            .overlapping(r.getStartKey() == null ? null : r.getStartKey().getRow(),
+                r.isStartKeyInclusive(), r.getEndKey() == null ? null : r.getEndKey().getRow())
+            .build();
```

Comment:
Always doing a metadata lookup per range for the batch scanner could be a performance problem.  Would be better to have some sort of caching.  Can cache previously seen extents as long as the cache is invalidated when scan server reports back that extent does not exists.

There may be some code from the bulk import code that we could use, for the case when bulk import works with an offline table.  I will look around for that.

Reply:
I found the KeyExtentCache in BulkImport. It looks like I should be able to reuse this, although I will likely have to move it to a different package. I may also need to modify it as it looks like the cached extents do not expire.

Reply:
@keith-turner - I created a new, different cache class in d73b879 that has different logic and implementation than the ConcurrentKeyExtentCache in the bulk import code. I'm going to kick off an IT run to ensure that this cache is working (I think it is based on the limited IT tests that I ran locally). Would be interested in your feedback on the logic in the new cache and whether or not it can replace the cache in the bulk import code.